Expand mapped tasks at DagRun.Veriy_integrity

ephraimbuddy opened the pull request:
Create the necessary task instances for a mapped task at dagrun.verify_integrity

Fixes https://github.com/apache/airflow/issues/22561

uranusjr commented on airflow/models/dagrun.py:
```diff
@@ -839,25 +839,59 @@ def task_filter(task: "Operator") -> bool:
+            def create_ti_mapping(task: "Operator", session: Session = session) -> List[dict]:
+
+                if isinstance(task, MappedOperator):
+                    indexes_to_map = task.get_indexes_to_map(self.run_id, session=session)
```
We can only ever be able to expand tasks with literal-only kwargs at this point, so we probably don’t need to call `_get_map_lengths` (which is expensive and accesses db), but instead rely solely on the task kwargs to get this length.

uranusjr commented on airflow/models/mappedoperator.py:
```diff
@@ -516,6 +515,34 @@ def _get_map_lengths(self, run_id: str, *, session: Session) -> Dict[str, int]:
+        from airflow.models.xcom_arg import XComArg
+
+        mapped_kwargs = self._get_expansion_kwargs()
+        map_lengths = [len(v) for _, v in mapped_kwargs.items() if not isinstance(v, XComArg)]
```
I’m thinking we probably should only expand here if _none of the mapped kwargs is XComArg_, because if there are XComArg, we very likely need to expand again later anyway.

ashb commented on airflow/models/dagrun.py:
```diff
@@ -839,25 +839,43 @@ def task_filter(task: "Operator") -> bool:
-                created_counts[task.task_type] += 1
-                return TI.insert_mapping(self.run_id, task, map_index=-1)
+            def create_ti_mapping(task: "Operator") -> List[dict]:
+                if isinstance(task, MappedOperator):
```
```suggestion
                if task.is_mapped:
```

(in a few places)

ephraimbuddy replied:
Should I move `is_mapped` to BaseOperator because mypy complains(on the code you gave me too @ashb which is same as the review above)

ashb replied:
Complains with what error specifically, because it is there https://github.com/apache/airflow/blob/436c17c655494eff5724df98d1a231ffa2142253/airflow/models/baseoperator.py#L1479

A useful thing: what type does mypy think `task` is here.
```suggestion
                reveal_type(task)
                if task.is_mapped:
```

And then run mypy, and it will print out info about `task` at mypy (this is not a real function, nothing is needed to be imported, it is just for mypy)

ephraimbuddy replied:
Have pushed the code. Revealed type is "Union[airflow.models.baseoperator.BaseOperator, airflow.models.mappedoperator.MappedOperator]". Quite confusing

ephraimbuddy replied:
I have seen why this is happening. `vars(mapped)` task returns an empty dict.

ashb replied:
That is because is_mapped is a property on the class (so it's not set in the `__dict__` of the object). What is the exact error mypy gives please?

ephraimbuddy replied:
Have type ignored it.
```
airflow/models/dagrun.py:828: error: Item "BaseOperator" of
"Union[BaseOperator, MappedOperator]" has no attribute
"mapped_args_are_literals"  [union-attr]
                if task.is_mapped and task.mapped_args_are_literals:
                                      ^
airflow/models/dagrun.py:830: error: Item "BaseOperator" of
"Union[BaseOperator, MappedOperator]" has no attribute "x"  [union-attr]
                    if ti.map_index >= task.x:
                                       ^
airflow/models/dagrun.py:870: note: Revealed type is "Union[airflow.models.baseoperator.BaseOperator, airflow.models.mappedoperator.MappedOperator]"
airflow/models/dagrun.py:871: error: Item "BaseOperator" of
"Union[BaseOperator, MappedOperator]" has no attribute
"mapped_args_are_literals"  [union-attr]
                if not task.is_mapped or not task.mapped_args_are_literals...
                                             ^
airflow/models/dagrun.py:873: error: Incompatible return value type (got
"Tuple[Union[BaseOperator, MappedOperator], range]", expected
"Tuple[Union[BaseOperator, MappedOperator], Tuple[int, ...]]")  [return-value]
                return (task, range(task.x))
                        ^
airflow/models/dagrun.py:873: error: Item "BaseOperator" of
"Union[BaseOperator, MappedOperator]" has no attribute "x"  [union-attr]
                return (task, range(task.x))
                                    ^
Found 5 errors in 1 file (checked 1 source file)

ERROR: The previous step completed with error. Please take a look at output above 
```

ashb commented on airflow/models/mappedoperator.py:
```diff
@@ -516,6 +515,34 @@ def _get_map_lengths(self, run_id: str, *, session: Session) -> Dict[str, int]:
+            for map_index in indexes_to_map:
+                ti: TaskInstance = TaskInstance(self, run_id=run_id, map_index=map_index)
+                tis.extend([ti])
+            return tis
```
Having to pass `mappings` in to here makes this function a lot more complex than it needs to be, and makes it very specific to the one place where it is currently called.

Instead I would have this function return the number of mapped tis to create (or maybe a cached property instead of a function?)

ephraimbuddy committed:
Expand mapped tasks at DagRun.Veriy_integrity

Create the necessary task instances for a mapped task at dagrun.verify_integrity

ashb reviewed (commented):
LGTM, but lets let @uranusjr take a look too.

uranusjr commented on airflow/models/mappedoperator.py:
```diff
@@ -677,3 +677,18 @@ def _find_index_for_this_field(index: int) -> int:
+        from airflow.models.xcom_arg import XComArg
+
+        expansion_kwargs = self._get_expansion_kwargs()
+        return sum(len(item) for item in expansion_kwargs.values() if not isinstance(item, XComArg))
```
This should probably be `literal_mapped_args_count_sum` since it does not count XComArg args. Also it may be a good idea to merge this with `mapped_args_are_literals`; both these are only called once and we shouldn’t need to iterate through the kwargs twice.

I’d do something higher level like

```python
def expand_ti_if_only_literal_mapped_args(self):
    expansion_kwargs = self._get_expansion_kwargs()
    literal_counts = [len(item) for item in expansion_kwargs.values() if isinstance(item, (list, dict))]
    if len(expansion_kwargs) != len(literal_counts):
        return [-1]  # Has non-literals, don't expand.
    # Move expand_mapped_literals here...
```

We should probably also refactor the `list, dict` part out into a `mappable_literal_types` constant (put it with `get_mappable_types`) so it’s easier to add support for more types in the future.

ashb replied:
Because of the use in the "check for removed or restored tasks" block earlier in `verify_integrity` I can't have the expansion here -- we need a count available to DagRun

uranusjr commented on airflow/models/dagrun.py:
```diff
@@ -810,20 +822,27 @@ def verify_integrity(self, session: Session = NEW_SESSION):
-                self.log.info("Restoring task '%s' which was previously removed from DAG '%s'", ti, dag)
-                Stats.incr(f"task_restored_to_dag.{dag.dag_id}", 1, 1)
-                ti.state = State.NONE
-            session.merge(ti)
```
This `session.merge()` call is gone in the new implementation; is it intended?

ashb replied:
Yes -- the TIs are fetched from the current attached session so this shouldn't be required?

uranusjr commented on airflow/models/dagrun.py:
```diff
@@ -810,20 +822,27 @@ def verify_integrity(self, session: Session = NEW_SESSION):
+                if ti.map_index >= task.mapped_args_count:  # type: ignore
+                    ti.state = State.REMOVED
+                elif ti.map_index < 0:
+                    ti.state = State.REMOVED
```
Too many `type: ignore` annotations, we should straighten out the types here.

ashb committed:
Refactor
